src/cargo/ops/cargo_rustc/context.rs

```diff
@@ -309,49 +309,55 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
 
     /// Get the metadata for a target in a specific profile
     pub fn target_metadata(&self, unit: &Unit) -> Option<Metadata> {
-        let metadata = unit.target.metadata();
+        // No metadata for dylibs because of a couple issues
+        // - OSX encodes the dylib name in the executable
+        // - Windows rustc multiple files of which we can't easily link all of them
+        if !unit.profile.test && unit.target.is_dylib() {
+            return None;
+        }
+
+        let metadata = unit.target.metadata().cloned().map(|mut m| {
+            if let Some(features) = self.resolve.features(unit.pkg.package_id()) {
+                let mut feat_vec: Vec<&String> = features.iter().collect();
+                feat_vec.sort();
+                for feat in feat_vec {
+                    m.mix(feat);
+                }
+            }
+            m.mix(unit.profile);
+            m
+        });
+        let mut pkg_metadata = {
+            let mut m = unit.pkg.generate_metadata();
+            if let Some(features) = self.resolve.features(unit.pkg.package_id()) {
+                let mut feat_vec: Vec<&String> = features.iter().collect();
+                feat_vec.sort();
+                for feat in feat_vec {
+                    m.mix(feat);
+                }
+            }
+            m.mix(unit.profile);
+            m
+        };
```

Comment:
In retrospect I'm unfortunately finding this kinda confusing (why we'd choose one or the other). I think nowadays it's probably best to just defer all of this until right here. That is, we should remove the `Metadata` inside of `Target` and calculate everything here (doing all the special casing that's happening during the construction. 

We seem to basically be doing all the special casing here anyway, so the initial construction in `toml.rs` isn't buying us much anyway.